### PR TITLE
Added retinaface-resnet50-pytorch model

### PR DESCRIPTION
**Pull request by Anna-Mironova**
- [x] Added GFLOPs and MParams

### CODE REVIEW

File: models/public/retinaface-resnet50-pytorch/README.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
+# retinaface-resnet50-pytorch
+
+## Use Case and High-Level Description
+
+The `retinaface-resnet50-pytorch` model is a PyTorch\* implementation of medium size model with ResNet50 backbone for Face Localization. It can output face bounding boxes and five facial landmarks in a single forward pass. More details provided in the [paper](https://arxiv.org/abs/1905.00641) and [repository](https://github.com/biubug6/Pytorch_Retinaface)
```
Comment (vladimir-dudnik) [inline models/public/retinaface-resnet50-pytorch/README.md:5]:
```suggestion
The `retinaface-resnet50-pytorch` model is a PyTorch\* implementation of medium size RetinaFace model with ResNet50 backbone for Face Localization. It can output face bounding boxes and five facial landmarks in a single forward pass. More details provided in the [paper](https://arxiv.org/abs/1905.00641) and [repository](https://github.com/biubug6/Pytorch_Retinaface)
```

File: models/public/retinaface-resnet50-pytorch/README.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+| GFLOPs                                                        | TBD         |
+| MParams                                                       | TBD         |
+| Source framework                                              | PyTorch\*   |
+
+Average Precision (AP) is defined as an area under the
+[precision/recall](https://en.wikipedia.org/wiki/Precision_and_recall)
+curve. All numbers were evaluated by taking into account only faces bigger than
+64 x 64 pixels.
+
+Accuracy validation approach different from described in the original repo.
```
Comment (vladimir-dudnik) [inline models/public/retinaface-resnet50-pytorch/README.md:21]: is it possible to briefly describe what are the differences and why they are needed?

Comment (eaidova) [reply models/public/retinaface-resnet50-pytorch/README.md:21]: original Wider evaluation protocol different from validation strategy in AC.
It splits whole validation set on 3 predefined subset (hard, medium, easy) and verify all these sets separately.

File: demos/common/python/models/retinaface.py [hunk 3]
```diff
@@ -302,3 +303,152 @@ def landmark_pred(boxes, landmark_deltas):
...
+            keep = nms(x_mins, y_mins, x_maxs, y_maxs, scores, self.nms_threshold,
+                       include_boundaries=not self._process_landmarks)
+
+            proposals = proposals[keep]
+            scores = scores[keep]
+            if self._process_landmarks:
+                landmarks = landmarks[keep]
+
+        result = []
+        if len(list(scores)) != 0:
```
Comment (eaidova) [inline demos/common/python/models/retinaface.py:385]:
```suggestion
        if np.size(scores):
```

File: demos/common/python/models/retinaface.py [hunk 4]
```diff
@@ -302,3 +303,152 @@ def landmark_pred(boxes, landmark_deltas):
...
+            if self._process_landmarks:
+                landmarks_x_coords = np.array(landmarks)[:, ::2] / scale_x
+                landmarks_y_coords = np.array(landmarks)[:, 1::2] / scale_y
+                for i in range(len(list(scores))):
+                    result.append(DetectionWithLandmarks(x_mins[i], y_mins[i], x_maxs[i], y_maxs[i], scores[i], 0,
+                                                         landmarks_x_coords[i], landmarks_y_coords[i]))
+            else:
+                for i in range(len(list(scores))):
+                    result.append(Detection(x_mins[i], y_mins[i], x_maxs[i], y_maxs[i], scores[i], 0))
+
```
Comment (eaidova) [inline demos/common/python/models/retinaface.py:403]:
```suggestion
                for i, s in enumerate(scores):
                    result.append(DetectionWithLandmarks(x_mins[i], y_mins[i], x_maxs[i], y_maxs[i], s, 0,
                                                         landmarks_x_coords[i], landmarks_y_coords[i]))
            else:
                for i, s in enumerate(scores):
                    result.append(Detection(x_mins[i], y_mins[i], x_maxs[i], y_maxs[i], s, 0))

```

Comment (eaidova) [reply demos/common/python/models/retinaface.py:403]: or even
```suggestion
                for x_min, y_min, x_max, y_max, score, landmarks_x, landmarks_y in zip(x_mins, y_mins, x_maxs, y_maxs, scores,  landmarks_x_coords, landmarks_y_coords):
                    result.append(DetectionWithLandmarks(x_min, y_min, x_max, y_max, score, 0,
                                                         landmarks_x, landmarks_y))
            else:
                for x_min, y_min, x_max, y_max, score in zip(x_mins, y_mins, x_maxs, y_maxs, scores)::
                    result.append(Detection(x_min, y_min, x_max, y_max, score, 0))

```

File: models/public/device_support.md [hunk 5]
```diff
@@ -123,6 +123,7 @@
 | resnet-50-tf | YES | YES | YES |
 | retinaface-anti-cov | YES |    |    |
 | retinaface-resnet50 | YES |    |    |
+| retinaface-resnet50-pytorch | YES |    |    |
```
Comment (eaidova) [inline models/public/device_support.md:126]: it really does not work on GPU or you did not verified?

File: models/public/retinaface-resnet50-pytorch/README.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+- `(x1, y1)` - coordinates of left eye
+- `(x2, y2)` - coordinates of rights eye
+- `(x3, y3)` - coordinates of nose
+- `(x4, y4)` - coordinates of left mouth corner
+- `(x5, y5)` - coordinates of right mouth corner
+
+### Converted model
+
+The converted model has the same parameters as the original model.
```
Comment (eaidova) [inline models/public/retinaface-resnet50-pytorch/README.md:84]:
```suggestion
The converted model has the same outputs as the original model.
```

File: models/public/retinaface-resnet50-pytorch/pre-convert.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+from torch import load, save
+
+from pathlib import Path
+
+
+def main():
```
Comment (eaidova) [inline models/public/retinaface-resnet50-pytorch/pre-convert.py:23]: I believe it can be done as part of model.py?
No need in additional script

File: demos/common/python/models/retinaface.py [hunk 8]
```diff
@@ -302,3 +303,153 @@ def landmark_pred(boxes, landmark_deltas):
...
+        self.n, self.c, self.h, self.w = self.net.input_info[self.image_blob_name].input_data.shape
+
+    def preprocess(self, inputs):
+        image = inputs
+
+        resized_image = resize_image(image, (self.w, self.h))
+        meta = {'original_shape': image.shape,
+                'resized_shape': resized_image.shape}
+        resized_image = resized_image.transpose((2, 0, 1))  # Change data layout from HWC to CHW
+        resized_image = resized_image.reshape((self.n, self.c, self.h, self.w))
```
Comment (eaidova) [inline demos/common/python/models/retinaface.py:334]: please use expand_dims instead (reshaping images to the same shape but with adding batch dimension looks reductant)

File: demos/object_detection_demo/python/object_detection_demo.py [hunk 9]
```diff
@@ -158,7 +159,8 @@ def __len__(self):
 def get_model(ie, args):
     input_transform = models.InputTransform(args.reverse_input_channels, args.mean_values, args.scale_values)
     common_args = (ie, args.model, input_transform)
-    if args.architecture_type in ('ctpn', 'yolo', 'yolov4', 'retinaface') and not input_transform.is_trivial:
+    if args.architecture_type in ('ctpn', 'yolo', 'yolov4', 'retinaface',
```
Comment (eaidova) [inline demos/object_detection_demo/python/object_detection_demo.py:162]: do we do not have such check in argparser?

File: models/public/retinaface-resnet50-pytorch/README.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+| GFLOPs                                                        | 88.8627     |
+| MParams                                                       | 27.2646     |
+| Source framework                                              | PyTorch\*   |
+
+Average Precision (AP) is defined as an area under the
+[precision/recall](https://en.wikipedia.org/wiki/Precision_and_recall)
+curve. All numbers were evaluated by taking into account only faces bigger than
+64 x 64 pixels.
+
+Accuracy validation approach different from described in the original [repository](https://github.com/biubug6/Pytorch_Retinaface/tree/master/widerface_evaluate). In contrast to the Accuracy Checker strategy where whole set is evaluated, the validation set is divided into 3 predefined subsets(hard, medium and easy) and all subsets are verified separately ij the original evaluation strategy.
```
Comment (eaidova) [inline models/public/retinaface-resnet50-pytorch/README.md:21]:
```suggestion
Accuracy validation approach different from described in the original [repository](https://github.com/biubug6/Pytorch_Retinaface/tree/master/widerface_evaluate). In contrast to the Accuracy Checker strategy where whole set is evaluated, the validation set is divided into 3 predefined subsets(hard, medium and easy) and all subsets are verified separately in the original evaluation strategy.
```

File: models/public/retinaface-resnet50-pytorch/model.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+from data.config import cfg_re50
+
+def remove_module_prefix(original_weights, converted_weights):
+    checkpoint = load(original_weights, map_location='cpu')
+    ckpt = {k.replace('module.', ''): v for k, v in checkpoint.items()}
+    save(ckpt, converted_weights)
+
+
+def create_model(original_weights, converted_weights):
+    remove_module_prefix(original_weights, converted_weights)
```
Comment (eaidova) [inline models/public/retinaface-resnet50-pytorch/model.py:27]: I mean slightly different.
You do not need rewrite state dict on the disk (just combain create model and load state dict in create model function with modification on the fly)

File: models/public/device_support.md [hunk 12]
```diff
@@ -123,6 +123,7 @@
 | resnet-50-tf | YES | YES | YES |
 | retinaface-anti-cov | YES |    |    |
 | retinaface-resnet50 | YES |    |    |
+| retinaface-resnet50-pytorch | YES | YES |    |
```
Comment (eaidova) [inline models/public/device_support.md:126]: BTW, please remove deleted models from this list

File: models/public/retinaface-resnet50-pytorch/pre-convert.py (hunk 7 above)
Comment (Anna-Mironova) [reply models/public/retinaface-resnet50-pytorch/pre-convert.py]: Done

File: demos/common/python/models/retinaface.py (hunk 8 above)
Comment (Anna-Mironova) [reply demos/common/python/models/retinaface.py]: Done

File: models/public/retinaface-resnet50-pytorch/README.md (hunk 10 above)
Comment (Anna-Mironova) [reply models/public/retinaface-resnet50-pytorch/README.md]: Done

File: models/public/retinaface-resnet50-pytorch/model.py (hunk 11 above)
Comment (Anna-Mironova) [reply models/public/retinaface-resnet50-pytorch/model.py]: Done

File: models/public/retinaface-resnet50-pytorch/README.md (hunk 1 above)
Comment (Anna-Mironova) [reply models/public/retinaface-resnet50-pytorch/README.md]: Done

File: models/public/retinaface-resnet50-pytorch/README.md (hunk 2 above)
Comment (Anna-Mironova) [reply models/public/retinaface-resnet50-pytorch/README.md]: Done

File: demos/object_detection_demo/python/README.md [hunk 13]
```diff
@@ -145,20 +147,22 @@ usage: object_detection_demo.py [-h] -m MODEL -at
...
   -m MODEL, --model MODEL
                         Required. Path to an .xml file with a trained model.
-  -at {ssd,yolo,yolov4,faceboxes,centernet,ctpn,retinaface,ultra_lightweight_face_detection}, --architecture_type {ssd,yolo,yolov4,faceboxes,centernet,ctpn,retinaface,ultra_lightweight_face_detection}
+  -at {ssd,yolo,yolov4,faceboxes,centernet,ctpn,retinaface,ultra_lightweight_face_detection,retinaface-pytorch}, --architecture_type {ssd,yolo,yolov4,faceboxes,centernet,ctpn,retinaface,ultra_lightweight_face_detection,retinaface-pytorch}
                         Required. Specify model' architecture type.
   -i INPUT, --input INPUT
                         Required. An input to process. The input must be a
-                        single image, a folder of images, video file or camera id.
+                        single image, a folder of images, video file or camera
+                        id.
```
Comment (vladimir-dudnik) [inline demos/object_detection_demo/python/README.md:165]: I'd better avoid word wrap for single quite short word

File: demos/object_detection_demo/python/README.md [hunk 14]
```diff
@@ -58,6 +58,12 @@ This file can be used as a parameter for [Model Downloader](../../../tools/downl
   - ctpn
 * architecture_type = faceboxes
   - faceboxes-pytorch
+* architecture_type = retinaface
+  - retinaface-anti-cov
+  - retinaface-resnet50
+  - ssh-mxnet
```
Comment (vladimir-dudnik) [inline demos/object_detection_demo/python/README.md:64]: these are missed models too?

**Review by vladimir-dudnik — COMMENTED** [review]
please review comments

**Review by eaidova — APPROVED** [review]

**Review by vladimir-dudnik — CHANGES_REQUESTED** [review]
please also review C++ demo readme.md and models.lst, it contains missed models too

let me know if C++ demo should be adopted for the new model

**Review by vladimir-dudnik — APPROVED** [review]
